### PR TITLE
Updating user attributes on changes in OAuth Provider

### PR DESCRIPTION
**Pull request by mpoornima**
#### Summary
Updates user's Username, Email, FirstName & LastName on login if the corresponding attributes have changed in the Auth Provider

#### Ticket Link
Jira: [PLT-3751](https://mattermost.atlassian.net/browse/PLT-3751)
GitHub: #5288   

#### Checklist
- [x] Added or updated unit tests
- [ ] Added API documentation
- [ ] Has UI changes
- [x] Includes text changes and localization file ([.../i18n/en.json](https://github.com/mattermost/platform/blob/master/i18n/en.json) and [.../webapp/i18n/en.json](https://github.com/mattermost/platform/tree/master/webapp/i18n/en.json)) updates
- [ ] Touches critical sections of the codebase (auth, upgrade, etc.)

Thanks

### CODE REVIEW

**Comment by jasonblais** [comment]
Thanks @mpoornima for the contribution!!

FYI, Linda will help with initial testing,

**Comment by lindalu** [comment]
Thanks @jasonblais, looks good to me! No issues found. Updates happened immediately and as expected.

Testing summary:
1. Signed up for account on http://i-0f2031b904a0fb470.spinmint.com:8065/pr5324 using GitLab OAuth
2. Verified full name, email, and username match GitLab settings.
3. Logged out of Mattermost.
4. Updated full name, email, and username in GitLab.
5. Logged into Mattermost using GitLab SSO.
6. Verified full name and username updated in Mattermost settings as expected.
7. Email did not update. 
8. Received email verification link from GitLab (aha!) and clicked link to verify new email address.
9. Logged out and back into Mattermost.
11. Verified email updated in Mattermost.

Additional testing:
1. Verified that changing username in GitLab to a username that already exists on that Mattermost server results in no change on either side.
2. Verified that changing email in GitLab to an email that already has an account on that Mattermost server results in no change on either side.

Note: In the Mattermost Account Settings, if the user tries to change name, username, or email, no edit fields are displayed, and messaging displays to let them know to do it in GitLab: "This field is handled through your login provider. If you want to change it, you need to do so through your login provider."

**Comment by coreyhulen** [comment]
@jwilander should also review.

File: api/user.go [hunk 1]
```diff
@@ -167,13 +167,61 @@ func LoginByOAuth(c *Context, w http.ResponseWriter, r *http.Request, service st
...
+
 	doLogin(c, w, r, user, "")
 	if c.Err != nil {
 		return nil
 	}
+
 	return user
 }
 
+func updateUserAttrs(userData io.Reader, user *model.User, provider einterfaces.OauthProvider, service string, siteURL string) *model.AppError {
```
Comment (jwilander) [inline api/user.go]: Would you be able to move this function into the `app` package in `app/user.go`?

Could you also update the function name to be `UpdateOAuthUserAttrs` so we know it's meant for OAuth?

File: api/user_test.go [hunk 2]
```diff
@@ -2608,3 +2612,176 @@ func TestGetByEmail(t *testing.T) {
 		}
 	}
 }
+
+func TestLoginByOAuthUpdateAttrs(t *testing.T) {
```
Comment (jwilander) [inline api/user_test.go]: Can you also move this test into the `app` package in `app/user_test.go`?

File: api/user.go (hunk 1 above)
Comment (jasonblais) [inline api/user.go]: @mpoornima Thanks for your contribution! 

Can you help review the feedback above?

Comment (mpoornima) [inline api/user.go]: Hey, I'm on vacation right now, will do this after I return (Feb 27th). Apologies for not notifying earlier.

Comment (jasonblais) [inline api/user.go]: Thanks @mpoornima for the update, no worries. Hope you have a great vacation! 

Looking forward to having this improvement in Mattermost after you come back :)

**Comment by mpoornima** [comment]
@jwilander made the requested changes. Please have a look.

**Comment by jasonblais** [comment]
This is good for 3.7, updated milestone. Thanks @mpoornima!

**Pull request merged**